### ISSUE TITLE
Clean up scheduler dependencies on k8s.io/kubernetes packages

### ISSUE THREAD

**Issue by damemi** [kubernetes/kubernetes#89930]
To make it easier for out-of-tree developers to import the scheduler framework, among other scheduling helpers and constants, we should explore areas that we can remove dependencies on `k8s.io/kubernetes`. Places that could be easily removed are, for example, using internal APIs where not required or importing unrelated packages for simple helpers which can be trivially implemented. This will clean up the dependency tree for importing scheduler code externally. 

This is related to: [Proposal to migrate scheduler code to staging](https://docs.google.com/document/d/1WO-ixERpqkCSEXEq30YtEH_z_G-BoLKeCbkRJcKq3xA/edit). While the proposal includes all scheduler code, we are only focusing on scheduler framework code. However identifying anywhere in the scheduler code that these dependencies can be cleaned up will be helpful as well.

This will probably be an ongoing cleanup effort, and breaking the work into smaller PRs may actually help with review and preventing code breaking in the process.
/kind cleanup
/sig scheduling

PRs opened so far:
* https://github.com/kubernetes/kubernetes/pull/90105 `pkg/apis/core` @xiaoanyunfei 
  * moving the default scheduler name and references to the external api
* ~~https://github.com/kubernetes/kubernetes/pull/90006~~ `pkg/api/core/v1/helper` @gavinfish (part 1)
  * copies `GetPersistentVolumeClaimClass`, which is used in several other areas around pkg/volume
  * replaced by https://github.com/kubernetes/kubernetes/pull/98433
* https://github.com/kubernetes/kubernetes/issues/91782 Overview issue of `pkg/api/core/v1/helper` @ingvagabund 
* ~~https://github.com/kubernetes/kubernetes/pull/90041 `pkg/api/testing` @tanjunchen~~
* https://github.com/kubernetes/kubernetes/pull/90039 ~~`pkg/scheduler/util`~~ `pkg/api/v1/pod` @yuzhiquan 🔹
  *  move `GetPodPriority` to scheduler utils
  * ~~https://github.com/kubernetes/kubernetes/pull/90703 `pkg/api/v1/pod` @yiyang5055 🔹~~
* https://github.com/kubernetes/kubernetes/pull/90046 `pkg/controller/volume/scheduling/metrics` @tanjunchen 
  * copies metrics registration and definitions around volume binding
* https://github.com/kubernetes/kubernetes/pull/90111 `pkg/features` @gavinfish 🔹
  * held on https://github.com/kubernetes/kubernetes/issues/90136 discussion on feature gating for staging components
* ~~https://github.com/kubernetes/kubernetes/pull/90000 `pkg/master/ports` @SataQiu~~
* ~~https://github.com/kubernetes/kubernetes/pull/90040~~ `pkg/util/node` @tanjunchen 🔹
  * copies Node util `GetZoneKey`, which should be moved to external and imported by both
  * replaced by https://github.com/kubernetes/kubernetes/pull/97818
* https://github.com/kubernetes/kubernetes/pull/90008 `pkg/util/parsers` @gavinfish 
  * removes `DefaultImageTag` (":latest") const only used by us and kubelet
* https://github.com/kubernetes/kubernetes/pull/90007 `pkg/volume/util` @gavinfish 
  * still have dependency on VolumeBinder

🔹 = Adds code to `pkg/scheduler/util`, which should be externalized

**Comment by damemi** [comment]
Just to note, another key part of this is not only cleaning up our usages but making sure that doesn't cause unnecessary duplication or interfere with other components.

For example, if we import constant `foo.Bar` and want to clean up the dependency by moving `Bar` to our package, we should check for other usages of `foo.Bar` in kubernetes code. If it's not used anywhere else we should delete it (no need to have it declared in 2 places), but if it is then the other usages should be taken into consideration (and perhaps also cleaned up as part of the PR)

This might require some collaboration with other owners, but will pay off in the long run to keep the code organized.

**Comment by YuikoTakada** [comment]
/assign
k8s.io/kubernetes/pkg/controller/volume/scheduling

**Comment by damemi** [comment]
I've updated the main description of this issue with a list of all the PRs we've received so far, if I missed any please let me know to add it.

We are trying to review them in a timely manner but please keep in mind that this is intended as an ongoing effort, and some package dependencies reveal trickier entanglement once you try to remove them so those may take longer to find a good resolution. Thanks to everyone that's submitted changes so far!

**Comment by damemi** [comment]
@ingvagabund has started moving some of these helpers to a newly-approved `library-go` staging repo, see https://github.com/kubernetes/kubernetes/pull/92507

**Comment by damemi** [comment]
Good news everyone! The new repo, [k8s.io/component-helpers](https://github.com/kubernetes/kubernetes/pull/92507), has been created with https://github.com/kubernetes/kubernetes/pull/92507 and is now open for relocating helper functions to be external. Huge thanks to @ingvagabund for taking on the work and feedback to get this repo initialized!

This has been the major blocker for much of the work that has already been started here, as many parts of the scheduler framework rely on internal helper functions that were also consumed by other components which we didn't want to duplicate.

If you have a PR that was already opened as part of this issue, please re-visit it with the new repo in mind! Refer to the [component-helpers README](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-helpers/README.md#component-helpers) as a guide for what may go in the repo.

I'll be going through and trying to revive some of the PRs originally linked to this issue that may have gone stale. Thanks everyone for your patience, and I apologize if you got stuck hanging on to a PR that didn't seem to go anywhere. We'll likely be able to unblock it now!

**Comment by ahg-g** [comment]
This is great, thanks for driving this @ingvagabund and @damemi

**Comment by gavinfish** [comment]
Thanks for @ingvagabund's great work!

And for this issue, @damemi I have some new question, when we change our dependencies from apis/core/helper to component-helpers, should we make the helper methods shared with other components, or we should just keep in scheduling?

**Comment by damemi** [comment]
@gavinfish we don't want to duplicate code, so if something is moving to the external repo it should move for all users

**Comment by damemi** [comment]
I discussed this issue in the sig-scheduling meeting today, but here is a comprehensive update on the current status of this.

Taking a more targeted approach to this, we want to better define the goals:
1. Make plugins independently importable by external projects (eg, Descheduler) by decoupling them from internal k8s.io/kubernetes packages
2. Similarly decouple scheduling framework dependencies to reduce the dependency tree for custom schedulers

It will be easiest to accomplish these in that order, so number 1 is to check for plugins that import `k8s.io/kubernetes`, but we don't necessarily care about framework dependencies yet (just for simplicity). In other words, `grep -rni "k8s.io/kubernetes" pkg/scheduler/framework/plugins/* | grep -v "k8s.io/kubernetes/pkg/scheduler"`

The following directories do not currently have any results, meaning these plugins at most only depend on scheduler packages:
```
pkg/scheduler/framework/plugins/defaultbinder
pkg/scheduler/framework/plugins/examples
pkg/scheduler/framework/plugins/imagelocality
pkg/scheduler/framework/plugins/interpodaffinity
pkg/scheduler/framework/plugins/nodelabel
pkg/scheduler/framework/plugins/nodename
pkg/scheduler/framework/plugins/nodeports
pkg/scheduler/framework/plugins/podtopologyspread
pkg/scheduler/framework/plugins/queuesort
pkg/scheduler/framework/plugins/serviceaffinity
pkg/scheduler/framework/plugins/volumerestrictions
```

The ones that do have the following imports:
```
pkg/scheduler/framework/plugins/defaultpreemption
  - k8s.io/kubernetes/pkg/controller/volume/scheduling
  - k8s.io/kubernetes/pkg/features

pkg/scheduler/framework/plugins/helper
  - k8s.io/kubernetes/pkg/apis/core (removed by https://github.com/kubernetes/kubernetes/pull/90105)

pkg/scheduler/framework/plugins/nodeaffinity
  - k8s.io/kubernetes/pkg/apis/core (removed by https://github.com/kubernetes/kubernetes/pull/90105)

pkg/scheduler/framework/plugins/nodepreferavoidpods
  - k8s.io/kubernetes/pkg/apis/core/v1/helper

pkg/scheduler/framework/plugins/noderesources
  - k8s.io/kubernetes/pkg/apis/core/v1/helper
  - k8s.io/kubernetes/pkg/features

pkg/scheduler/framework/plugins/nodeunschedulable
  - k8s.io/kubernetes/pkg/apis/core/v1/helper (removed by https://github.com/kubernetes/kubernetes/pull/98445)

pkg/scheduler/framework/plugins/nodevolumelimits
  - k8s.io/kubernetes/pkg/apis/core/v1/helper
  - k8s.io/kubernetes/pkg/volume/util
  - k8s.io/kubernetes/pkg/features

pkg/scheduler/framework/plugins/selectorspread
  - k8s.io/kubernetes/pkg/util/node

pkg/scheduler/framework/plugins/tainttoleration
  - k8s.io/kubernetes/pkg/apis/core/v1/helper (removed by https://github.com/kubernetes/kubernetes/pull/98445)

pkg/scheduler/framework/plugins/volumebinding
  - k8s.io/kubernetes/pkg/controller/volume/scheduling
  - k8s.io/kubernetes/pkg/features
  - k8s.io/kubernetes/pkg/controller/volume/persistentvolume/util

pkg/scheduler/framework/plugins/volumezone
  - k8s.io/kubernetes/pkg/apis/core/v1/helper (removed by https://github.com/kubernetes/kubernetes/pull/98433)
```

These plugins should be specifically worked on to reduce these dependencies. For example:
* API helpers should be reviewed to determine if they specifically belong under API approval, or are simply performing common functions on API objects and can be moved to a sig-owned [component-helpers](https://github.com/kubernetes/component-helpers) directory
* Featuregating logic should be refactored to not directly depend on `pkg/features`, but consider alternatives such as adding fields to the internal plugin argument types, which can be passed feature gate flags during initialization (need to discuss an exact approach for this, but there are options)

It's possible that not every plugin can feasibly make these changes, but the above list should give anyone still interested in contributing to this issue a good starting point.

**Comment by ahg-g** [comment]
A great deal of progress have been made on this already, can we get an update on where we stand now? what is left?

**Comment by kerthcet** [comment]
/remove-lifecycle rotten

Just want to shout out where are we now? Is there anything left? cc @damemi

**Comment by damemi** [comment]
@kerthcet this has fallen off my radar unfortunately. There's definitely more work to be done, though I don't know any specific examples right now. If you would like to take the lead on reviving it please feel free :)

**Comment by kerthcet** [comment]
/remove-lifecycle stale

I'll revisit this once I have time, this has been opened for quite a long time. Or we can open a new ticket, quite a long conversation here.

**Comment by sanposhiho** [comment]
/unassign @YuikoTakada
Let me know if you're still working on, I believe not so unassign for now.

/remove-help
Will add again if needed. 

---

/cc @macsko @dom4ha 

Revisiting here since I had a conversation around the scheduling framework's Golang breaking changes [recently](https://github.com/kubernetes/kubernetes/pull/129126#issuecomment-2533649294). 

So, there are two motivations and we can think them separately:
1. move the scheduling framework outside to help custom plugin developers.
2. move in-tree plugin implementation outside to help descheduler (or other people who have to import in-tree plugins).
  a. @damemi just to confirm the current status, does the descheduler team still want it?

```
$ grep -rni "k8s.io/kubernetes" pkg/scheduler/framework/* | grep -v "k8s.io/kubernetes/pkg/scheduler" | grep -v plugins
pkg/scheduler/framework/events.go:25:   "k8s.io/kubernetes/pkg/features"
pkg/scheduler/framework/events_test.go:31:      "k8s.io/kubernetes/pkg/features"
pkg/scheduler/framework/preemption/preemption.go:40:    apipod "k8s.io/kubernetes/pkg/api/v1/pod"
pkg/scheduler/framework/runtime/framework.go:42:        "k8s.io/kubernetes/pkg/util/slice"
pkg/scheduler/framework/types.go:38:    "k8s.io/kubernetes/pkg/features"
pkg/scheduler/framework/types_test.go:35:       "k8s.io/kubernetes/pkg/features"
pkg/scheduler/framework/types_test.go:38:       "k8s.io/kubernetes/test/utils/ktesting"
pkg/scheduler/framework/types_test.go:39:       "k8s.io/kubernetes/test/utils/ktesting/initoption"
```

```
$ grep -rni "k8s.io/kubernetes" pkg/scheduler/framework/* | grep -v "k8s.io/kubernetes/pkg/scheduler" | grep plugins   
pkg/scheduler/framework/plugins/nodename/node_name_test.go:27:  "k8s.io/kubernetes/test/utils/ktesting"
pkg/scheduler/framework/plugins/nodevolumelimits/utils.go:27:   "k8s.io/kubernetes/pkg/features"
pkg/scheduler/framework/plugins/nodevolumelimits/csi_test.go:38:        "k8s.io/kubernetes/test/utils/ktesting"
pkg/scheduler/framework/plugins/registry.go:21: "k8s.io/kubernetes/pkg/features"
pkg/scheduler/framework/plugins/schedulinggates/scheduling_gates_test.go:29:    "k8s.io/kubernetes/test/utils/ktesting"
pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go:29:   v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
pkg/scheduler/framework/plugins/noderesources/fit.go:31:        v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
pkg/scheduler/framework/plugins/noderesources/resource_allocation.go:28:        "k8s.io/kubernetes/pkg/features"
pkg/scheduler/framework/plugins/dynamicresources/dynamicresources_test.go:46:   "k8s.io/kubernetes/test/utils/ktesting"
pkg/scheduler/framework/plugins/nodeunschedulable/node_unschedulable_test.go:27:        "k8s.io/kubernetes/test/utils/ktesting"
pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:50:        apipod "k8s.io/kubernetes/pkg/api/v1/pod"
pkg/scheduler/framework/plugins/volumebinding/binder_test.go:46:        "k8s.io/kubernetes/pkg/controller"
pkg/scheduler/framework/plugins/volumebinding/binder_test.go:47:        pvtesting "k8s.io/kubernetes/pkg/controller/volume/persistentvolume/testing"
pkg/scheduler/framework/plugins/volumebinding/binder.go:47:     v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
pkg/scheduler/framework/plugins/volumebinding/binder.go:48:     "k8s.io/kubernetes/pkg/features"
```

**Comment by sanposhiho** [comment]
And, among those two, let's aim at the easier one, the framework staging repo, first. At a glance, the biggest blocker was `k8s.io/kubernetes/pkg/features`. I saw some previous discussion ([example](https://github.com/kubernetes/kubernetes/issues/90136)), but looks like no progress and that'd be a time-consuming blocker.

However, I'd like to propose a different angle here; _only exposing types and not exposing implementations_. The use case for the scheduling framework staging repo would be mainly custom plugin developers.
And, the custom plugin developers only need to import types (e.g., each interfaces, `NodeInfo`, `PodInfo`, etc), but they don't need an implementation. For example, `pkg/scheduler/framework/types.go` depends on the feature flag for the resource calculations for `NodeInfo`, but users don't need to import this calculation part actually.
If we go with this way, I believe we can start to craft the scheduling framework staging repo even now.

---

Also, I know this strategy wouldn't work for (2) in-tree plugin implementation staging repo because the implementation is what descheduler is needing, not only types. So, we, either way, need to deal with the feature-flag dependency problem for (2).

**Comment by sanposhiho** [comment]
And, among those two, let's aim at the easier one, the framework staging repo, first. At a glance, the biggest blocker was `k8s.io/kubernetes/pkg/features`. I saw some previous discussion ([example](https://github.com/kubernetes/kubernetes/issues/90136)), but looks like no progress and that'd be a time-consuming blocker.

However, I'd like to propose a different angle here; _only exposing types and not exposing implementations_. The use case for the scheduling framework staging repo would be mainly custom plugin developers.
And, the custom plugin developers only need to import types (e.g., each interfaces, `NodeInfo`, `PodInfo`, etc), but they don't need an implementation. For example, `pkg/scheduler/framework/types.go` depends on the feature flag for the resource calculations for `NodeInfo`, but users don't need to import this calculation part actually.
If we go with this way, I believe we can start to craft the scheduling framework staging repo even now. What does everyone think? cc @alculquicondor @dom4ha @macsko 

---

Also, I know this strategy wouldn't work for (2) in-tree plugin implementation staging repo because the implementation is what descheduler is needing, not only types. So, we, either way, need to deal with the feature-flag dependency problem for (2).

**Comment by sanposhiho** [comment]
@macsko @dom4ha @alculquicondor Any thoughts on my proposal? In short, I'm proposing starting to move the scheduling framework types (e.g. plugin interfaces etc) to the staging repo to try to eliminate the major motivation of users to import k/k, which is for custom plugins.

**Comment by macsko** [comment]
> I'm proposing starting to move the scheduling framework types (e.g. plugin interfaces etc) to the staging repo so that custom plugin developers would no longer have to import k/k.

IMO moving the interfaces seems reasonable. Is there a drawback?

**Comment by sanposhiho** [comment]
Yeah, I don't see any downside to move only the interfaces and types, not everything.

**Comment by sanposhiho** [comment]
@alculquicondor What would be a next step if we proceed creating a staging repo? Would it need a KEP, or can we just go ahead?

**Comment by alculquicondor** [comment]
Here are the instructions https://github.com/kubernetes/kubernetes/tree/master/staging#creating-a-new-repository-in-staging

However, is it really worth it at the moment? We haven't heard much from the descheduler team.

We know the DRA folks are coming with extensive features with tight deadlines, so I'm not sure it's the best use of our time to pursue this right now.

**Comment by sanposhiho** [comment]
As I [classified](https://github.com/kubernetes/kubernetes/issues/89930#issuecomment-2566236328), exporting the framework is the motivation for custom plugin developers, which is separate from descheduler's motivation which wants plugins to be on the staging repo.

I've seen arguments/complains around this problem sometimes both online and offline (e.g., [the latest one](https://github.com/kubernetes/kubernetes/pull/129126#issuecomment-2533649294)); basically importing k/k isn't something we can recommend in general and hence we have to move, at least, the framework to eliminate the need of importing k/k for plugin developers. 

Also, again, moving plugins to the staging repo is another story and we may not have to, depending on how much descheduler team need now.

> We know the DRA folks are coming with extensive features with tight deadlines, so I'm not sure it's the best use of our time to pursue this right now.

Yeah, I totally agree to potentially skip this depending on the remaining bandwidth of us. It's something we have to at some point, but not necessarily right now urgently.

**Comment by sanposhiho** [comment]
I'd like to start this (scope: only interface and types) towards the next release. I don't want to keep the current situation, keep putting the breaking changes on every scheduler type/interface on the action required section of kubernetes releases.

cc @macsko @dom4ha

**Comment by sanposhiho** [comment]
Looks like https://github.com/kubernetes/kubernetes/tree/master/staging#creating-a-new-repository-in-staging doesn't require KEP, we can just go ahead moving things away from k/k scheduler. _Maybe_, we need a blog post or something to announce it at the release though, that's another thing that we can consider later.

**Comment by sanposhiho** [comment]
It would be awesome if anyone else could help here (I can work on it on my own if no one, though)

**Comment by dom4ha** [comment]
What do you think about discussing the scope and implementation details in the form of KEP?

**Comment by sanposhiho** [comment]
It'd be nicer to have a doc to make sure the current scope and a possible future expansion so that everyone would be on the same page before starting. 
However, I'm not sure KEP would be the most suitable..? We don't need a feature gate for this change (that'd be very hard: switching the package's import path based on the feature gate...? how?) and also don't need a feature stage (alpha/beta/ga) too. So, e.g., just a Google doc might be sufficient enough?

**Comment by ania-borowiec** [comment]
Sounds good! I'll start working on this early next week.

**Comment by utam0k** [comment]
> However, I'm not sure KEP would be the most suitable..? We don't need a feature gate for this change (that'd be very hard: switching the package's import path based on the feature gate...? how?) and also don't need a feature stage (alpha/beta/ga) too. So, e.g., just a Google doc might be sufficient enough?

This change is crucial for users who create custom scheduler plugins. So, would we be able to write the Kubernetes blog?
And, also we need to update the plugins in scheduler-plugins...	I'll support this migration and write a blog if needed.

**Comment by sanposhiho** [comment]
We internally discussed and, for now, we're making the migration process clearer on the release note, not the blog. Like we'll edit the release note directly for the guidance of migration, which we believe should be enough because basically it's just import path changes.

**Comment by ania-borowiec** [comment]
I'm happy to let you know that Plugin and Handle interfaces have been moved to k8s.io/kube-scheduler together with all relevant types. That means that out-of-tree plugins should not need to import kubernetes/kubernetes anymore.

**Comment by sanposhiho** [comment]
Awesome! Thanks for all the long effort here, Ania!

Do we agree that we can close this issue now? I think there is another huge topic, whether we want to move our in-tree plugins to the staging repo, but we haven't heard anything recently from the descheduler team here. So, maybe the motivation is not that big today. Or, even if it _is_ still relevant, probably it's nicer to create another issue anyway because the discussion here is very long.

**Comment by macsko** [comment]
Agree, let's close this issue and open a new one, when further movement would need to be considered.

**Comment by sanposhiho** [comment]
/close

🎉

**Comment by sanposhiho** [comment]
I'll make another issue with the summary of the related discussion here

**Comment by sanposhiho** [comment]
https://github.com/kubernetes/kubernetes/issues/133994
